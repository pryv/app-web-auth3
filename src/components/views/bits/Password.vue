<template>
  <div>
    <v-text-field
      :id="id"
      :value="value"
      :appendIcon="visiblePass ? 'lock_open' : 'lock'"
      :type="visiblePass ? 'text' : 'password'"
      :rules="[rules.required]"
      :label="label"
      @input="$emit('input', $event)"
      @click:append="() => (visiblePass = !visiblePass)"/>

    <v-text-field
      v-if="confirmation"
      :id="`${id}Confirmation`"
      v-model="repass"
      :type="visiblePass ? 'text' : 'password'"
      :rules="[matchPassword]"
      :label="`${label} confirmation`"/>
  </div>
</template>

<script>
export default {
  props: {
    confirmation: { type: Boolean, default: false },
    value: { type: String, default: '' },
    label: { type: String, default: 'Password' },
    id: { type: String, default: 'password' },
  },
  data: () => ({
    visiblePass: false,
    repass: '',
    rules: {
      required: value => !!value || 'Password is required.',
    },
  }),
  computed: {
    matchPassword: function () {
      return this.value === this.repass || 'Password confirmation does not match.';
    },
  },
};
</script>
