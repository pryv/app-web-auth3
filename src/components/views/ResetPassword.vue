<template>
  <div>
    <h1>{{ pageTitle }}</h1>

    <v-form
      v-if="showForm"
      ref="form"
      v-model="validForm"
      @submit.prevent>

      <v-text-field
        id="usernameOrEmail"
        v-model="ctx.user.username"
        :rules="[rules.required]"
        label="Username or email"/>

      <Password
        v-if="resetToken!=null"
        v-model="password"
        :confirmation="true"/>

      <v-btn
        id="submitButton"
        :disabled="!validForm||submitting"
        @click="submit"
      >{{ buttonText }}</v-btn>
    </v-form>
    <!--
    // If the reset password happend during an Auth Proceess (pollUrl exists)
    -->
    <div v-if="ctx.isAccessRequest()">
      <v-divider class="mt-3 mb-2"/>
      <router-link :to="{ name: 'Authorization' }"><h3>Go to Sign in</h3></router-link>
    </div>

    <Alerts
      :successMsg="success"
      :errorMsg="error"/>
  </div>
</template>

<script>
import Password from './bits/Password';
import Alerts from './bits/Alerts';
import Context from '../../context.js';
import controllerFactory from '../controller/controller.js';

export default {
  components: {
    Password,
    Alerts,
  },
  props: {
    resetToken: { type: String, default: null },
  },
  data: () => ({
    password: '',
    error: '',
    success: '',
    showForm: true,
    submitting: false,
    ctx: {},
    c: null,
    rules: {
      required: value => !!value || 'This field is required.',
    },
    validForm: false,
  }),
  computed: {
    pageTitle: function () {
      return this.resetToken ? 'Set a new password' : 'Reset password';
    },
    buttonText: function () {
      return this.resetToken ? 'Change password' : 'Request password reset';
    },
  },
  async created () {
    this.ctx = new Context(this.$route.query);
    await this.ctx.init();
    this.c = controllerFactory(this.ctx);
  },
  methods: {
    submit () {
      this.ctx.user.username = this.ctx.user.username.trim();
      if (this.$refs.form.validate()) {
        this.submitting = true;
        // Ask for a reset token
        if (this.resetToken == null) {
          this.c.requestResetPassword()
            .then(() => {
              this.showForm = false;
              this.success = 'We have sent password reset instructions to your e-mail address.';
            })
            .catch(this.showError)
            .finally(() => { this.submitting = false; });
        } else {
          // If we already got a reset token, we can change the password
          this.c.resetPassword(this.password, this.resetToken)
            .then(() => {
              this.showForm = false;
              this.success = 'Your password have been successfully changed.';
            })
            .catch(this.showError)
            .finally(() => { this.submitting = false; });
        }
      }
    },
    showError (error) {
      this.error = error.msg;
    },
  },
};
</script>
