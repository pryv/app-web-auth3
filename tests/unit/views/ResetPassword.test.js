import { shallowMount } from '@vue/test-utils';
import ResetPassword from '@/components/views/ResetPassword';
import Password from '@/components/views/bits/Password';
import Vue from 'vue';
import Vuetify from 'vuetify';

Vue.use(Vuetify);

describe('ResetPassword.test.js', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallowMount(ResetPassword, {
      mocks: {
        $route: {query: {}},
        $t: (key) => key,
      },
      stubs: ['router-link'],
    });
  });

  it('renders correctly (snapshots matching)', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it('deactivates submit button when form is not yet filled', () => {
    const submitButton = wrapper.find('#submitButton');
    expect(wrapper.vm.validForm).toBe(false);
    expect(submitButton.attributes().disabled).toBe('true');
  });

  it('activates submit button when form is valid', () => {
    wrapper.setData({
      validForm: true,
    });
    const submitButton = wrapper.find('#submitButton');
    expect(submitButton.attributes().disabled).toBeFalsy();
  });

  it('validates required fields correctly', () => {
    const requiredRule = wrapper.vm.rules.required;
    expect(requiredRule()).toBe('This field is required');
    expect(requiredRule('notEmpty')).toBe(true);
  });

  it('shows password reset form by default', () => {
    // Renders password reset form
    const pageTitle = wrapper.find('h1').text();
    expect(pageTitle).toBe('Reset password');
    const submitButton = wrapper.find('#submitButton');
    expect(submitButton.text()).toBe('Reset');
    const usernameOrEmail = wrapper.find('#usernameOrEmail');
    expect(usernameOrEmail.exists()).toBe(true);
    // Does not ask for password
    const password = wrapper.find(Password);
    expect(password.exists()).toBe(false);
  });

  it('shows password change form when resetToken is provided', () => {
    wrapper.setProps({
      resetToken: 'reset',
    });
    // Renders password change form
    const pageTitle = wrapper.find('h1').text();
    expect(pageTitle).toBe('Set a new password');
    const submitButton = wrapper.find('#submitButton');
    expect(submitButton.text()).toBe('Change');
    const usernameOrEmail = wrapper.find('#usernameOrEmail');
    expect(usernameOrEmail.exists()).toBe(true);
    // Ask for password
    const password = wrapper.find(Password);
    expect(password.exists()).toBe(true);
  });
});
