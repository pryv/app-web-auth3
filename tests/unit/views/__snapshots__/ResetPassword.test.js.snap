// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResetPassword.test.js renders correctly (snapshots matching) 1`] = `
<div>
  <h1>
    Reset password
  </h1>
   
  <v-form-stub>
    <v-text-field-stub
      id="usernameOrEmail"
      label="Username or email"
      rules="function required(value) {
          return !!value || 'This field is required.';
        }"
      value=""
    />
     
    <!---->
     
    <v-btn-stub
      activeclass="v-btn--active"
      disabled="true"
      id="submitButton"
      tag="button"
      type="button"
    >
      Request password reset
    </v-btn-stub>
  </v-form-stub>
   
  <!---->
   
  <alerts-stub
    errormsg=""
    successmsg=""
  />
</div>
`;
